Tools/LyTestTools/ly_test_tools/o3de/editor_test.py

```diff
@@ -0,0 +1,777 @@
+import ly_test_tools.environment.waiter as waiter
+import ly_test_tools.environment.process_utils as process_utils
+
+from ly_test_tools.o3de.asset_processor import AssetProcessor
+from ly_test_tools.launchers.exceptions import WaitTimeoutError
+from . import editor_test_utils as editor_utils
+
+# This file provides editor testing functionality to easily write automated editor tests for O3DE.
+# For using these utilities, you can subclass your test suite from EditorTestSuite, this allows an easy way of specifying 
+# python test scripts that the editor will run without needing to write any boilerplace code.
+# It supports out of the box parallelization(running multiple editor instances at once), batching(running multiple tests in the same editor instance) and
+# crash detection.
+# Usage example:
+#    class MyTestSuite(EditorTestSuite):
+#   
+#        class MyFirstTest(EditorSingleTest):
+#            from . import script_to_be_run_by_editor as test_module
+#   
+#        class MyTestInParallel_1(EditorParallelTest):
+#            from . import another_script_to_be_run_by_editor as test_module
+#        
+#        class MyTestInParallel_2(EditorParallelTest):
+#            from . import yet_another_script_to_be_run_by_editor as test_module
+#
+#
+# EditorTestSuite does introspection of the defined classes inside of it and automatically prepares the tests, parallelizing/batching as required
+
+# This file contains no tests, but with this we make sure it won't be picked up by the runner since the file ends with _test
+__test__ = False
+
+# Abstract base class for an editor test.
+class EditorTestBase(ABC):
+    # Maximum time for run, in seconds
+    timeout = 180
+    # Test file that this test will run
+    test_module = None
+
+# Test that will be run alone in one editor
+class EditorSingleTest(EditorTestBase):
+    # Extra cmdline arguments to supply to the editor for the test
+    extra_cmdline_args = []
+
+    # Custom setup function, will run before the test
+    @staticmethod
+    def setup(instance, request, workspace, editor, editor_test_results, launcher_platform):
+        pass
+
+    # Custom run wrapping. The code before yield will run before the test, and after the yield after the test
+    @staticmethod
+    def wrap_run(instance, request, workspace, editor, editor_test_results, launcher_platform):
+        yield
+
+    # Custom teardown function, will run after the test    
+    @staticmethod
+    def teardown(instance, request, workspace, editor, editor_test_results, launcher_platform):
+        pass
+
+# Test that will be both be run in parallel and batched with eachother in a single editor.
+# Does not support per test setup/teardown for avoiding any possible race conditions
+class EditorSharedTest(EditorTestBase):
+    # Specifies if the test can be batched in the same editor
+    is_batchable = True
+    # Specifies if the test can be run in multiple editors in parallel
+    is_parallelizable = True
```

Comment:
Wonder if this should be explicitly named ```EditorBatchedParallelTest``` (or something similar) instead of equating ```Shared``` to mean Batched and Parallel.
Thoughts?

Reply:
I was thinking on that, however I wanted to provide a common parent class for doing the search. 
Initially we only had SharedTest but wanted to provide a way to easily specify parallel only or batch only